File: chor-extraction/src/main/java/ast/sp/nodes/Offering.java
package ast.sp.nodes;

import ast.sp.interfaces.Behaviour;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by lara on 04/04/17.
 */
public class Offering implements Behaviour {
    private final String process;
    private final HashMap<String, Behaviour> labels;

    public Offering(String process, HashMap<String, Behaviour> labels) {
        this.process = process;
        this.labels = labels;
    }

    public String getProcess() {
        return process;
    }

    public HashMap<String, Behaviour> getLabels() {
        return labels;
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append(process + "&{");
        for (Map.Entry<String, Behaviour> entry : labels.entrySet()) {
            builder.append(entry.getKey() + ": " + entry.getValue() + ", ");
        }
        builder.delete(builder.length()-2, builder.length());
        builder.append("}");
        return builder.toString();
    }
}

File: chor-extraction/src/main/java/ast/sp/nodes/Offering.kt
package ast.sp.nodes

import ast.sp.interfaces.Behaviour
import ast.sp.interfaces.Interaction
import java.util.*

data class Offering(val process: String, val labels: HashMap<String, Behaviour>) : Interaction {
    override fun toString(): String {
        val builder = StringBuilder()
        builder.append(process + "&{")
        for ((key, value) in labels) {
            builder.append("$key: $value, ")
        }
        builder.delete(builder.length - 2, builder.length)
        builder.append("}")
        return builder.toString()
    }

    override fun findRecProcCall(procname: String): Boolean {
        for ((key, value) in labels) {
            if (value.findRecProcCall(procname)) {
                return true
            }
        }
        return false
    }

    override fun copy(): Interaction {
        val tempmap = labels.clone()
        return Offering(process, tempmap as HashMap<String, Behaviour>)
    }

    override fun equals(b: Behaviour): Boolean {
        if (!(b is Offering) || process != b.process) return false
        else {
            for (label in labels) {
                val bl = b.labels.get(label.key)
                if (bl == null || !label.value.equals(bl)) return false
            }
        }
        return true
    }
}
